airflow/api_fastapi/core_api/routes/ui/grid.py

```diff
@@ -0,0 +1,324 @@
+from airflow.utils.task_group import MappedTaskGroup, TaskGroup
+
+grid_router = AirflowRouter(prefix="/grid", tags=["Grid"])
+
+
+@grid_router.get(
+    "/{dag_id}",
+    include_in_schema=False,
+    responses=create_openapi_http_exception_doc([status.HTTP_400_BAD_REQUEST, status.HTTP_404_NOT_FOUND]),
+)
+def grid_data(
+    dag_id: str,
+    run_types: QueryDagRunRunTypesFilter,
+    run_states: QueryDagRunStateFilter,
+    session: SessionDep,
+    offset: QueryOffset,
+    request: Request,
+    num_runs: QueryLimit,
+    base_date: OptionalDateTimeQuery = None,
+    root: str | None = None,
+    filter_upstream: bool = False,
+    filter_downstream: bool = False,
+) -> GridResponse:
+    """Return grid data."""
+    ## Database calls to retrieve the DAG Runs and Task Instances and validate the data
+    dag: DAG = request.app.state.dag_bag.get_dag(dag_id)
+    if not dag:
+        raise HTTPException(status.HTTP_404_NOT_FOUND, f"Dag with id {dag_id} was not found")
+
+    if root:
+        dag = dag.partial_subset(
+            task_ids_or_regex=root, include_upstream=filter_upstream, include_downstream=filter_downstream
+        )
+
+    current_time = timezone.utcnow()
+    # Retrieve, sort and encode the previous DAG Runs
+    base_query = (
+        select(
+            DagRun.run_id,
+            DagRun.queued_at,
+            DagRun.start_date,
+            DagRun.end_date,
+            DagRun.state,
+            DagRun.run_type,
+            DagRun.data_interval_start,
+            DagRun.data_interval_end,
+            DagRun.dag_version_id.label("version_number"),
+        )
+        .select_from(DagRun)
+        .where(DagRun.dag_id == dag.dag_id, DagRun.logical_date <= func.coalesce(base_date, current_time))
+        .order_by(DagRun.id.desc())
+    )
+
+    def get_dag_run_sort_param():
+        """Get the Sort Param for the DAG Run."""
+
+        def _get_run_ordering_expr(name: str) -> ColumnOperators:
+            """Get the Run Ordering Expression."""
+            expr = DagRun.__mapper__.columns[name]
+            # Data interval columns are NULL for runs created before 2.3, but SQL's
+            # NULL-sorting logic would make those old runs always appear first. In a
+            # perfect world we'd want to sort by ``get_run_data_interval()``, but that's
+            # not efficient, so instead the columns are coalesced into logical_date,
+            # which is good enough in most cases.
+            if name in ("data_interval_start", "data_interval_end"):
+                expr = func.coalesce(expr, DagRun.logical_date)
+            return expr.desc()
+
+        ordering_expression = (_get_run_ordering_expr(name) for name in dag.timetable.run_ordering)
+        # create SortParam with ordering_expression and DagRun.id.desc()
+        return ordering_expression
+
+    dag_runs_select_filter, _ = paginated_select(
+        statement=base_query.order_by(*get_dag_run_sort_param(), DagRun.id.desc()),
```

Comment:
Because manually handling the sorting will yield things like that. By default things are sorted `asc` but then you need to muanually add the secondary sort on the primary key.

And we do not handle desc and asc sorting. Only forced `desc`

Reply:
I agree, I converted this to SortParam creation as a logic